src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketSCMSourceListener.java

```diff
@@ -0,0 +1,89 @@
+package com.atlassian.bitbucket.jenkins.internal.scm;
+
+import hudson.Extension;
+import hudson.model.listeners.ItemListener;
+import java.util.List;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+import jenkins.model.Jenkins;
+import jenkins.scm.api.SCMSourceOwner;
+
+@Extension
+public class BitbucketSCMSourceListener extends ItemListener {
+    // Set constants for thread management
+    // Setting max threads to 20 to avoid overwhelming the Bitbucket API.
+    private static final int MAX_THREADS = 20;
+    private static final int MIN_THREADS = 1;
+    private static final int THREAD_CPU_FACTOR = 4;
+    private static final Logger LOGGER = Logger.getLogger(BitbucketSCMSourceListener.class.getName());
+
+    /**
+     * This method is called when the Jenkins instance is loaded. It initializes all
+     * BitbucketSCMSource instances found in the Jenkins instance.
+     *
+     * @see ItemListener#onLoaded() for more details.
+     */
+    @Override
+    public void onLoaded() {
+        // Get the Jenkins instance and all items that implement SCMSourceOwner
+        Jenkins j = Jenkins.get();
+        List<SCMSourceOwner> owners = j.getAllItems(SCMSourceOwner.class);
+
+        // Filter the owners to find those that have BitbucketSCMSource instances
+        List<BitbucketSCMSource> sources =
```

Comment:
This is an improvement over readResolve as each folder/job is processed sequentially in readResolve

Reply:
Should we make the number of threads configurable via the Jenkins config/system property or environment variables?

Reply:
In testing this took init times down from 15 minutes to 45 seconds with 2000 jobs. It also doesn't block Jenkins from starting up and loads the jobs in the background as expected.